benchmarks/bench_plot_randomized_svd.py

```diff
@@ -0,0 +1,448 @@
+benchmarks with `enable_spectral_norm = False`, as Frobenius' is MUCH faster to
+compute.
+
+The benchmarks follow.
+
+(a) plot: time vs norm, varying number of power iterations
+    data: many datasets
+    goal: compare normalization policies and study how the number of power
+    iterations affect time and norm
+
+(b) plot: n_iter vs norm, varying rank of data and number of components for
+    randomized_SVD
+    data: low-rank matrices on which we control the rank
+    goal: study whether the rank of the matrix and the number of components
+    extracted by randomized SVD affect "the optimal" number of power iterations
+
+(c) plot: time vs norm, varing datasets
+    data: many datasets
+    goal: compare default configurations
+
+We compare the following algorithms:
+-   randomized_svd(..., power_iteration_normalizer='none')
+-   randomized_svd(..., power_iteration_normalizer='LU')
+-   randomized_svd(..., power_iteration_normalizer='QR')
+-   randomized_svd(..., power_iteration_normalizer='auto')
+-   fbpca.pca() from https://github.com/facebook/fbpca (if installed)
+
+Conclusion
+----------
+- n_iter=2 appears to be a good default value
+- power_iteration_normalizer='none' is OK if n_iter is small, otherwise LU
+  gives similar errors to QR but is cheaper. That's what 'auto' implements.
+
+References
+----------
+(1) Finding structure with randomness: Stochastic algorithms for constructing
+    approximate matrix decompositions
+    Halko, et al., 2009 http://arxiv.org/abs/arXiv:0909.4061
+
+(2) A randomized algorithm for the decomposition of matrices
+    Per-Gunnar Martinsson, Vladimir Rokhlin and Mark Tygert
+
+(3) An implementation of a randomized algorithm for principal component
+    analysis
+    A. Szlam et al. 2014
+"""
+
+# Author: Giorgio Patrini
+
+import numpy as np
+import scipy as sp
+import matplotlib.pyplot as plt
+
+import gc
+import pickle
+from time import time
+from collections import defaultdict
+import os.path
+
+from sklearn.utils import gen_batches
+from sklearn.utils.validation import check_random_state
+from sklearn.utils.extmath import randomized_svd
+from sklearn.datasets.samples_generator import (make_low_rank_matrix,
+                                                make_sparse_uncorrelated)
+from sklearn.datasets import (fetch_lfw_people,
+                              fetch_mldata,
+                              fetch_20newsgroups_vectorized,
+                              fetch_olivetti_faces,
+                              fetch_rcv1)
+
+try:
+    import fbpca
+    fbpca_available = True
+except ImportError:
+    fbpca_available = False
+
+# If this is enabled, tests are much slower and will crash with the large data
+enable_spectral_norm = False
+
+# Determine when to switch to batch computation for matrix norms,
+# in case the reconstructed (dense) matrix is too large
+MAX_MEMORY = np.int(2e9)
+
+CIFAR_FOLDER = "../../scikit_learn_data/cifar-10-batches-py/"
+SVHN_FOLDER = "../../scikit_learn_data/SVHN/"
```

Comment:
As those datasets are not part of the datasets that cannot be loaded via the scikit-learn datasets helper, I would not expect to fetch them from there. Instead I would just do:

```python
# The following dataset can be dowloaded manually from:
# URL CIFAR 10
# URL SVHN
CIFAR_FOLDER = "./scikit_learn_data/cifar-10-batches-py/"
SVHN_FOLDER = "./SVHN/"
```

FYI, it's possible to lookup the pass to the `scikit_learn_data` folder with:

```python
>>> from sklearn.datasets import get_data_home
>>> get_data_home()
```

At some point it would be great to refactor the CIFAR / SVHN loaders as generic loaders to be part of `sklearn.datasets.fetch_*` but this can be done in a later PR.